pkg/instancetype/compatibility.go

```diff
@@ -11,18 +11,29 @@ import (
 	instancetypev1alpha2 "kubevirt.io/api/instancetype/v1alpha2"
 )
 
-func decodeOldObject(data []byte, isPreference bool) (runtime.Object, error) {
-	if isPreference {
-		oldObject, err := decodeOldPreferenceRevisionObject(data)
-		if err != nil {
-			return nil, fmt.Errorf("failed to decode old revision object: %w", err)
-		}
-		if oldObject == nil {
-			return nil, nil
-		}
-		return convertPreference(oldObject)
+func isInstancetypeSpec(data []byte) bool {
+	return json.Unmarshal(data, &instancetypev1alpha1.VirtualMachineInstancetypeSpec{}) == nil
+}
+
+func isPreferenceSpec(data []byte) bool {
+	return json.Unmarshal(data, &instancetypev1alpha1.VirtualMachinePreferenceSpec{}) == nil
+}
+
+func decodeOldObject(data []byte) (runtime.Object, error) {
```

Comment:
This method does not work as expected. 

The `data` variable does not contain JSON serialization of `instancetypev1alpha1.VirtualMachineInstancetypeSpec` or `instancetypev1alpha1.VirtualMachinePreferenceSpec`. It contains serialization of `instancetypev1alpha1.VirtualMachineInstancetypeSpecRevision` or `instancetypev1alpha1.VirtualMachinePreferenceSpecRevision`.
But we cannot differentiate between them based on the success of `json.Unmarshal()` call, because both of those structures have the same definition, so it is possible to unmarshal one into the other.

In general, I don't like replacing a boolean parameter with an expensive operation like JSON decoding.

Reply:
ACK thanks, I assumed we had coverage somewhere for this but apparently not. I've dropped this rework from the initial refactor change and kept the boolean in place.